### ISSUE TITLE
Simplify LTO flags?

### ISSUE THREAD

**Issue by kripken** [emscripten-core/emscripten#10324]
Currently we have a bunch of LTO flags and LTO-affecting flags, like `--llvm-lto`, `-flto`, `-s WASM_OBJECT_FILES=0`. Talking this over with @dschuff , perhaps we can simplify all this to making the standard `-flto` do everything, that is, that flag would

 * Make us use bitcode in object files (during compilation)
 * Make us use bitcode in system libraries (during link)
 * Run LTO optimizations (during link)

We could then remove the other nonstandard flags. That would be a breaking change but a simple one, in that the old flags would error + point to the proper standard flag `-flto`, so just a minor one-time annoyance for users.

One significant change is that this would make `-flto` use bitcode system libraries, which is a change for people previously using `-flto`. Also we would no longer have a way to do LTO on the user's code but not system libraries. Do we need that?

(This might best wait for after removing fastcomp, as it would be easier to do then.)

**Comment by sbc100** [comment]
Related perhaps to: https://github.com/emscripten-core/emscripten/pull/7961

**Comment by sbc100** [comment]
With was partially fixed in #10490

**Comment by sbc100** [comment]
Do we want to keep `--llvm-lto` around?

In general with the llvm backend, if you give bitcode files to the linker that is a sign that you want to do LTO.  The is an equivalent flag to wasm-ld (and lld in general) which as `-lto-O0`, `-lto-O1`, etc.

This defaults to level 2, since it assumes that anyone feeding bitcode to the linker probably wants some LTO.   Note that this does *not* effect the codegen opt level which (I think) is mostly embedded in each bitcode file. 

I propose that we follow LLVM and default `-lto-O2` by not passing any arguments by default.

The question is then, do we do anything at all with `--lvm-lto` ?  The choices I see are:

1. Convert it to -lto-ON
2. Ignore it.
3. Error out saying its a legacy option.

I'm tempted to do (1) at least initially.

**Comment by dschuff** [comment]
Agreed that by default, feeding bitcode to the linker means you want LTO, so we should default to lto-O2. My preference would be to do option 1 for now as well. But if the default is to optimize all bitcode inputs with O2, wouldn't 1 be equivalent to 2 anyway?

**Comment by dschuff** [comment]
Oh, nevermind, I forgot that `-llvm-lto` takes an argument. I think I agree with @kripken in  #10606. We can just ignore `-llvm-lto`, default to O2 with bitcode, and if the user really does want to change that, they can still use `-Wl,-lto-O3`, right?

**Comment by sbc100** [comment]
Exactly.  Agreed.

**Comment by dschuff** [comment]
Is it the case today that if you happen to feed a bitcode file to the linker, it will automatically run LTO optimizations? Will it include bitcode system libraries?

**Comment by sbc100** [comment]
> Is it the case today that if you happen to feed a bitcode file to the linker, it will automatically run LTO optimizations? Will it include bitcode system libraries?

Any bitcode you feed into the linker will be LTO'd yes.  This is the behaviour of lld generally and there is not way to disable this.

The only way (today) to get bitcode system libraries is to pass `-flto` at link time the compiler driver.

**Comment by dschuff** [comment]
ah ok, that makes sense. I hadn't realized that there was any additional effect of `-flto` at link time. (Is this documented anywhere btw?).
One thing that made me think about this topic (even before the stale-bot woke this thread up) is that a user with a big app recently noted that their link time was really slow, and I noticed that some of their linker inputs had '.bc' suffixes and I was wondering exactly how much that was affecting their link time, and if there was any good way for them to figure out in more detail what is going on.

**Comment by sbc100** [comment]
One problem is that `.bc` suffix doesn't mean bitcode contents.  A lot of project for historical reasons use `.bc` suffix without `-flto` so end up with object files ... perhaps we should add a warning about this:

`warning: compiling it files with ".bc" extension but without using `-flto` will resulting is a normal object file and not a bitcode file.   Suggest avoiding the use of ".bc" for object files, or adding "-flto"`   ?

**Comment by dschuff** [comment]
Maybe. In general I think I prefer compilers to be less opinionated about suffixes than emscripten even is today. Build systems certainly do random things sometimes. It's likely that such a warning might mean build system annoyance to fix, and I might be annoyed at getting warned about something that was working perfectly fine. 
OTOH given that a file suffix with .bc is likely to be emscripten-specific already, maybe it's not so bad?
IMO the surprising thing is that handing bitcode to the linker results in optimization, but maybe outside emscripten bitcode files are rare enough that it's not a big deal.

Maybe the problem is if a user is doing a link with O0 or no optimization, but still feeding bitcode to the linker. Linker opt flags aren't as much of a thing outside emscripten, but are pretty essential with emscripten. Maybe we should warn when linking without opts (indicating the user might want faster linking) but linking a bitcode file. Would that be costly or annoying to implement in emcc? I guess we'd also want a way to disable it; disabling warnings at link time is also kind of weird.

**Comment by sbc100** [comment]
I think we just want to match behaviour of native `lld` which I think we already do.  

If you feed bitcode objects to the linker there is no way to avoid at least link time compilation, right?

I think the opt level that is used using LTO is at least partly driven by the bitfile file anyway.. if you build a bitcode file with `-O0` then it will embed `optnone` in the bitcode for each function.. and the linker will obey this.  So you will get the fastest possible "LTO" in that case.

Why would somebody compile with `-flto` if they want fast link times?  Have you seen folks doing that?   Are they compiling with `-flto + -O2` and still expecting fast link times?

**Comment by dschuff** [comment]
Oh, that's a good point, I forgot that optlevel was embedded in the functions now. I agree that what we have now is probably the best we can get, since obviously it at least has to be compiled before linking.

In the case I saw, I didn't see all the compile command lines, I just saw someone posted their linker command and some files had a .bc extension. It could have been that they were really just .o files (perhaps they had been using fastcomp before and they were bc files then), or maybe some part of that user's build was done in a different way, or something.
